### PR TITLE
Add com.ticktick.TickTick

### PR DESCRIPTION
**Pull request by gkeep**
### Please confirm your submission meets all the criteria

- [x] I have read the [App Requirements][reqs] and [App Maintenance][maint] pages.
- [x] My pull request follows the instructions at [App Submission][submission].
- [x] I am using only the minimal set of permissions. *(If not, please explain each non-standard permission.)*
- [x] All assets referenced in the manifest are redistributable by any party.  If not, the unredistributable parts are using an extra-data source type.
- [ ] I am an upstream contributor to the project. If not, I contacted upstream developers about submitting their software to Flathub. **Link:**
- [ ] I own the domain used in the application ID or the domain has a policy for delegating subdomains (e.g. GitHub, SourceForge).
- [x] Any additional patches or files have been submitted to the upstream projects concerned. *(If not, explain why.)*

[reqs]: https://github.com/flathub/flathub/wiki/App-Requirements
[maint]: https://github.com/flathub/flathub/wiki/App-Maintenance
[submission]: https://github.com/flathub/flathub/wiki/App-Submission

I will be contacting the upstream developers soon.

### CODE REVIEW

**Comment by gkeep** [comment]
I have created a ticket upstream. Will update here once/if I get a response.

**Review by hfiguiere — COMMENTED** [review]
Since you unpack an x86_64 binary you need to ad this 

https://github.com/flathub/flathub/wiki/App-Maintenance#only-build-on-x86_64

or add support for aarch64

File: com.ticktick.TickTick.yml [hunk 1]
```diff
@@ -0,0 +1,59 @@
...
+        --strip-components=3 \
+        ./opt/TickTick
+      - |
+        bsdtar -Oxf ticktick.deb 'data.tar*' |
+        bsdtar -xf - \
+        -C ${FLATPAK_DEST}/ \
+        --strip-components=2 \
+        ./usr
+
+      - install -Dm755 ${FLATPAK_DEST}/TickTick/ticktick -t ${FLATPAK_DEST}/bin/
```
Comment (hfiguiere) [inline com.ticktick.TickTick.yml:46]: you should symlink instead

Comment (hfiguiere) [reply com.ticktick.TickTick.yml:46]: or move. But this is duplicating the content.

**Comment by gkeep** [comment]
> support for aarch64

Please provide any documentation or instructions on this, I couldn't find anything. TickTick provides an arm64 .deb, so I can unpack the binary the same way as with x86_64.

File: com.ticktick.TickTick.yml (hunk 1 above)
Comment (gkeep) [reply com.ticktick.TickTick.yml:46]: Since a wrapper script is actually executed by the .desktop file this is completely redundant.

**Comment by gkeep** [comment]
I've limited build arches to x86_64 for now.

**Comment by hfiguiere** [comment]
bot, build com.ticktick.TickTick

**Comment by hfiguiere** [comment]
> > support for aarch64
> 
> Please provide any documentation or instructions on this, I couldn't find anything. TickTick provides an arm64 .deb, so I can unpack the binary the same way as with x86_64.

You can mark a source with "only-arches" to indicate it's specific to an arch. See https://docs.flatpak.org/en/latest/flatpak-builder-command-reference.html

**Comment by hfiguiere** [comment]
It's missing the AppStream file.

**Comment by gkeep** [comment]
bot, build com.ticktick.TickTick

File: com.ticktick.TickTick.yml [hunk 2]
```diff
@@ -0,0 +1,58 @@
...
+        --strip-components=2 \
+        ./usr
+
+      - desktop-file-edit --set-key=Exec --set-value=ticktick ${FLATPAK_DEST}/share/applications/ticktick.desktop
+      - install -Dm755 ticktick -t ${FLATPAK_DEST}/bin/
+
+    sources:
+      - type: file
+        dest-filename: ticktick.deb
+        url: https://appest-public.s3.amazonaws.com/download/linux/linux_deb_x64/ticktick-0.0.5-amd64.deb
```
Comment (barthalion) [inline com.ticktick.TickTick.yml:52]: You should add a stanza for https://github.com/flathub/flatpak-external-data-checker/ to automate updates.

File: com.ticktick.TickTick.yml [hunk 3]
```diff
@@ -0,0 +1,58 @@
...
+        bsdtar -xf - \
+        -C ${FLATPAK_DEST}/ \
+        --strip-components=2 \
+        ./usr
+
+      - desktop-file-edit --set-key=Exec --set-value=ticktick ${FLATPAK_DEST}/share/applications/ticktick.desktop
+      - install -Dm755 ticktick -t ${FLATPAK_DEST}/bin/
+
+    sources:
+      - type: file
```
Comment (barthalion) [inline com.ticktick.TickTick.yml:50]: This should be `extra-data` as this is a proprietary application we have no rights to redistribute.

File: com.ticktick.TickTick.metainfo.xml [hunk 4]
```diff
@@ -0,0 +1,32 @@
...
+<component type="desktop">
+  <id>com.ticktick.Ticktick</id>
+  <launchable type="desktop-id">com.ticktick.Ticktick.desktop</launchable>
+  <name>TickTick</name>
+  <project_license>LicenseRef-proprietary</project_license>
+  <developer_name>TickTick Team</developer_name>
+  <summary>TickTick: Todo list, checklist and task manager app</summary>
+  <description>
+    <p>Join millions of people to capture ideas, organize to-dos, and make the most of your life with TickTick.</p>
+    <p>Available for free, syncs across iOS, Android, Mac, Windows, Linux and Web.</p>
```
Comment (barthalion) [inline com.ticktick.TickTick.metainfo.xml:11]: This should include a disclaimer like https://flathub.org/apps/details/com.discordapp.Discord does.

File: com.ticktick.TickTick.metainfo.xml [hunk 5]
```diff
@@ -0,0 +1,32 @@
+<?xml version="1.0" encoding="UTF-8"?>
+<component type="desktop">
+  <id>com.ticktick.Ticktick</id>
+  <launchable type="desktop-id">com.ticktick.Ticktick.desktop</launchable>
+  <name>TickTick</name>
+  <project_license>LicenseRef-proprietary</project_license>
+  <developer_name>TickTick Team</developer_name>
+  <summary>TickTick: Todo list, checklist and task manager app</summary>
```
Comment (barthalion) [inline com.ticktick.TickTick.metainfo.xml:8]: There's no need to repeat application name in the summary field.

File: com.ticktick.TickTick.metainfo.xml [hunk 6]
```diff
@@ -0,0 +1,32 @@
...
+  <description>
+    <p>Join millions of people to capture ideas, organize to-dos, and make the most of your life with TickTick.</p>
```
Comment (barthalion) [inline com.ticktick.TickTick.metainfo.xml:10]: This blurb should be generally extended with information from https://www.ticktick.com/.

**Comment by gkeep** [comment]
Do I need to install the AppStream metainfo somehow, since the build fails?

**Comment by gkeep** [comment]
Can't seem to make it work with `type: extra-data`, because it is downloaded during installation and not building. Please help.

File: flathub.json [hunk 7]
```diff
@@ -1,3 +1,3 @@
 {
-  "only-arches": ["x86_64"]
+  "only-arches": ["x86_64", "aarch64"]
```
Comment (hfiguiere) [inline flathub.json:2]: you can remove this file, it's not needed. these are the only two arches supported.

**Comment by gkeep** [comment]
I get this error now when trying to build:
```
Error: icon ticktick not found below $dir/.flatpak-builder/rofiles/rofiles-iA73YJ/files/share/icons
```

I already install icons to `${FLATPAK_DEST}/share/icons/hicolor/${icon}x${icon}/apps/$FLATPAK_ID.png` and I have no idea what am I doing wrong. Any pointers?

File: com.ticktick.TickTick.yml [hunk 8]
```diff
@@ -0,0 +1,77 @@
+app-id: com.ticktick.TickTick
+runtime: org.freedesktop.Platform
+runtime-version: '21.08'
+sdk: org.freedesktop.Sdk
+base: org.electronjs.Electron2.BaseApp
+base-version: '21.08'
+
+command: ticktick
+rename-desktop-file: ticktick.desktop
+rename-icon: ticktick
```
Comment (hfiguiere) [inline com.ticktick.TickTick.yml:10]: you install the icons with the right name.
```suggestion
```

But then you need to change the icon name in the `.desktop`

File: com.ticktick.TickTick.yml [hunk 9]
```diff
@@ -0,0 +1,77 @@
...
+  - --talk-name=org.kde.StatusNotifierWatcher
+
+modules:
+  - name: ticktick
+    buildsystem: simple
+    build-commands:
+      - install -Dm755 apply_extra /app/bin/apply_extra
+      - install -Dm755 ticktick -t ${FLATPAK_DEST}/bin/
+      - install -Dm644 ticktick.desktop -t ${FLATPAK_DEST}/share/applications 
+      - desktop-file-edit --set-key=Exec --set-value=ticktick ${FLATPAK_DEST}/share/applications/ticktick.desktop
```
Comment (hfiguiere) [inline com.ticktick.TickTick.yml:33]: since the .deskop is in the repository, you might as well make it right and withou the need to use this.

**Comment by gkeep** [comment]
bot, build com.ticktick.TickTick

**Comment by gkeep** [comment]
bot, build com.ticktick.TickTick

File: com.ticktick.TickTick.metainfo.xml [hunk 10]
```diff
@@ -0,0 +1,46 @@
+<?xml version="1.0" encoding="UTF-8"?>
+<component type="desktop">
+  <id>com.ticktick.Ticktick</id>
```
Comment (hfiguiere) [inline com.ticktick.TickTick.metainfo.xml:3]:
```suggestion
  <id>com.ticktick.TickTick</id>
```

it's case sensitive

File: com.ticktick.TickTick.metainfo.xml [hunk 11]
```diff
@@ -0,0 +1,46 @@
...
+  <launchable type="desktop-id">com.ticktick.Ticktick.desktop</launchable>
```
Comment (hfiguiere) [inline com.ticktick.TickTick.metainfo.xml:4]:
```suggestion
  <launchable type="desktop-id">com.ticktick.TickTick.desktop</launchable>
```
case sensitive

File: com.ticktick.TickTick.desktop [hunk 12]
```diff
@@ -0,0 +1,9 @@
+[Desktop Entry]
+Name=TickTick
+Exec=/opt/TickTick/ticktick %U
```
Comment (hfiguiere) [inline com.ticktick.TickTick.desktop:3]:
```suggestion
Exec=ticktick %U
```

File: com.ticktick.TickTick.metainfo.xml (hunk 11 above)
Comment (gkeep) [reply com.ticktick.TickTick.metainfo.xml:4]: Oh wow, didn't even notice those.

File: com.ticktick.TickTick.yml [hunk 13]
```diff
@@ -0,0 +1,77 @@
...
+  - --talk-name=org.kde.StatusNotifierWatcher
+
+modules:
+  - name: ticktick
+    buildsystem: simple
+    build-commands:
+      - install -Dm755 apply_extra /app/bin/apply_extra
+      - install -Dm755 ticktick -t ${FLATPAK_DEST}/bin
+      - install -Dm644 $FLATPAK_ID.desktop -t ${FLATPAK_DEST}/share/applications 
+      - desktop-file-edit --set-key=Exec --set-value=ticktick ${FLATPAK_DEST}/share/applications/$FLATPAK_ID.desktop
```
Comment (hfiguiere) [inline com.ticktick.TickTick.yml:31]:
```suggestion
```

you can remove this with the suggested change in the `.desktop`

**Comment by gkeep** [comment]
bot, build com.ticktick.TickTick

**Comment by gkeep** [comment]
bot, build com.ticktick.TickTick

**Comment by gkeep** [comment]
Thanks for all the help!

File: com.ticktick.TickTick.yml (hunk 2 above)
Comment (gkeep) [reply com.ticktick.TickTick.yml:49]: Currently, there is no way to check for a new version using any of the methods.

**Comment by gkeep** [comment]
Response for the ticket:
>Hi there,
>
>Thanks for your feedback.
>I will note it down as feature request and bring it up to the product team for future consideration.
>I'll definitely let you know if there be any plans in the future.

**Comment by gkeep** [comment]
Everything works good, apart from Wayland support (I guess they're using an old version of Electron). Merge?

**Comment by barthalion** [comment]
bot, build com.ticktick.TickTick

**Pull request closed without merging**